keras_hub/src/models/qwen_moe/qwen_moe_decoder.py

```diff
@@ -0,0 +1,636 @@
+            num_experts=self.num_experts,
+            hidden_dim=self.hidden_dim,
+            intermediate_dim=self.intermediate_dim,
+            kernel_initializer=self.kernel_initializer,
+            name="experts",
+            dtype=self.dtype_policy,
+        )
+        self.expert_bank.build(decoder_sequence_shape)
+
+        self.shared_expert_dense = QwenMoeMLP(
+            intermediate_dim=self.intermediate_dim_shared,
+            hidden_dim=self.hidden_dim,
+            kernel_initializer=self.kernel_initializer,
+            layer_norm_epsilon=self.layer_norm_epsilon,
+            name="shared_expert_dense",
+            dtype=self.dtype_policy,
+        )
+        self.shared_expert_dense.build(decoder_sequence_shape)
+
+        self.shared_expert_gate_dense = keras.layers.Dense(
+            1,
+            use_bias=False,
+            name="shared_expert_gate_dense",
+            dtype=self.dtype_policy,
+        )
+        self.shared_expert_gate_dense.build(decoder_sequence_shape)
+
+        self.built = True
+
+    def call(self, hidden_states, attention_mask=None, training=None):
+        batch_size, seq_len, _ = ops.shape(hidden_states)
+        hidden_states_flattened = ops.reshape(
+            hidden_states, (-1, self.hidden_dim)
+        )
+
+        router_logits = self._sparse_feedforward_gate_dense(
+            hidden_states_flattened
+        )
+        router_probs = ops.softmax(router_logits, axis=-1)
+
+        top_p, top_i = ops.top_k(router_probs, k=self.top_k)
+        if self.norm_topk_prob:
+            top_p = top_p / ops.sum(top_p, axis=-1, keepdims=True)
+
+        one_hot = ops.one_hot(top_i, self.num_experts)
+        one_hot = ops.cast(one_hot, top_p.dtype)
+        routing_full = ops.sum(one_hot * top_p[..., None], axis=1)
+        routing_full = ops.transpose(routing_full, (1, 0))
+        routing_full = ops.cast(routing_full, hidden_states_flattened.dtype)
+
+        expert_out = self.expert_bank(hidden_states_flattened)
+
+        weighted_out = expert_out * routing_full[:, :, None]
+        expert_contribution = ops.sum(weighted_out, axis=0)
+
+        shared_expert_output = self.shared_expert_dense(hidden_states_flattened)
+        shared_expert_output *= ops.sigmoid(
+            self.shared_expert_gate_dense(hidden_states_flattened)
+        )
+
+        out_flat = expert_contribution + shared_expert_output
+        out = ops.reshape(out_flat, (batch_size, seq_len, self.hidden_dim))
+
+        # Compute and add auxiliary loss during training
+        if training:
+            aux_loss = compute_load_balancing_loss(
+                router_logits=router_logits,
+                num_experts=self.num_experts,
+                top_k=self.top_k,
+                attention_mask=attention_mask,
+            )
+            self.add_loss(self.router_aux_loss_coef * aux_loss)
+
+        return out, router_logits
+
+
+class QwenMoeTransformerDecoder(keras.layers.Layer):
```

Comment:
add a self.run_layer_test for the decoder

Reply:
i just noticed that we don't have layer tests for other model? Why is that the case?

Reply:
Ahh, sorry only needed for exported layers - no need to add them for these